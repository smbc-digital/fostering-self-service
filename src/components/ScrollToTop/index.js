import { useEffect } from 'react'
import { withRouter } from 'react-router'
import PropTypes from 'prop-types'

export const ScrollToTop = ({ location, children }) => {
	useEffect(() => {
		window.scrollTo(0, 0)
	}, [location])

	return children
}

ScrollToTop.propTypes = {
	location: PropTypes.object,
	children: PropTypes.object,
}

export default withRouter(ScrollToTop)
