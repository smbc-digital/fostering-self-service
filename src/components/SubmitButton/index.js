import React, { Fragment, useState } from 'react'
import PropTypes from 'prop-types'
import { Button, Anchor } from 'smbc-react-components'
import { Applicant } from 'constants'

const SubmitButton = ({ currentApplicant, secondApplicant, onSaveAndGoBackClick, history, isValid, isLoading }) => {
    const [saveAndGoBackClicked, setSaveAndGoBackClicked] = useState(false)

    if (currentApplicant === Applicant.FirstApplicant && secondApplicant) {
        return <Fragment>
            <Button
                label="Next step"
                isValid={isValid}
                isLoading={isLoading}
            />
            <Anchor
                label='Previous'
                history={history} />
        </Fragment>
    }

    return <Fragment>
        <Button
            label="Save and next step"
            isValid={!saveAndGoBackClicked && isValid}
            isLoading={isLoading && !saveAndGoBackClicked}
        />
        <Anchor
            label='Previous'
            history={history} />
        <Button
            label="Save and go back to fostering area"
            isValid={!isLoading && !saveAndGoBackClicked && isValid}
            isLoading={isLoading && saveAndGoBackClicked}
            colour='inverted'
            onButtonClick={event => {
                    setSaveAndGoBackClicked(true)
                    onSaveAndGoBackClick(event)
                }
            }
            useLeftChevron={true} />
    </Fragment>
}

SubmitButton.propTypes = {
    currentApplicant: PropTypes.string,
    secondApplicant: PropTypes.object,
    history: PropTypes.object,
    onSaveAndGoBackClick: PropTypes.func,
    isValid: PropTypes.bool,
    isLoading: PropTypes.bool
}

SubmitButton.defaultProps = {
    isLoading: false,
    isValid: true
}

export default SubmitButton