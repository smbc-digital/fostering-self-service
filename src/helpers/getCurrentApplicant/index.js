import { Applicant, SECOND_APPLICANT_ROUTE } from 'constants'

const getCurrentApplicant = match => {
    if (match.params && match.params[0] === SECOND_APPLICANT_ROUTE) {
        return Applicant.SecondApplicant
    }

    return Applicant.FirstApplicant
}

export default getCurrentApplicant