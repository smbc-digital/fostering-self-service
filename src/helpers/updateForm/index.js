import { fetchWithTimeout } from '../index'

const TaskStatus = {
    None: 0,
    Completed: 1,
    NotCompleted: 2,
    CantStart: 3
}

const StageName = {
    HomeVisit: 0,
    Application: 1
}

const HomeVisitFormName = {
    TellUsAboutYourself: 0,
    YourEmploymentDetails: 1,
    LanguagesSpokenInYourHome: 2,
    YourPartnership: 3,
    YourFosteringHistory: 4,
    YourHealth: 5,
    TellUsAboutYourInterestInFostering: 6,
    YourHousehold: 7,
    ChildrenLivingAwayFromYourHome: 8
}

const ApplicationFormName = {
    References: 0,
    GpDetails: 1,
    AddressHistory: 2
}

const getHomeVisitUpdateEndpoint = form => {
    switch (form) {
        case HomeVisitFormName.TellUsAboutYourself:
            return '/fostering/home-visit/about-yourself'
        case HomeVisitFormName.YourEmploymentDetails:
            return '/fostering/home-visit/your-employment-details'
        case HomeVisitFormName.LanguagesSpokenInYourHome:
            return '/fostering/home-visit/languages-spoken-in-your-home'
        case HomeVisitFormName.YourFosteringHistory:
            return '/fostering/home-visit/your-fostering-history'
        case HomeVisitFormName.YourPartnership:
            return '/fostering/home-visit/partnership-status'
        case HomeVisitFormName.TellUsAboutYourInterestInFostering:
            return '/fostering/home-visit/interest-in-fostering'
        case HomeVisitFormName.YourHealth:
            return '/fostering/home-visit/about-your-health'
        case HomeVisitFormName.YourHousehold:
            return '/fostering/home-visit/household'
        case HomeVisitFormName.ChildrenLivingAwayFromYourHome:
            return '/fostering/home-visit/children-living-away-from-home'
        default:
            throw new Error('No matching endpoint for given form.')
    }
}

const getApplicationUpdateEndpoint = form => {
    switch (form) {
        case ApplicationFormName.References:
            return '/fostering/application/references'
        case ApplicationFormName.GpDetails:
            return '/fostering/application/gp-details'
        default:
            throw new Error('No matching endpoint for given form.')
    }
}

const getFormStageEndpoint = stage => {
    switch (stage) {
        case StageName.HomeVisit: 
            return '/fostering/home-visit'
        case StageName.Application:
            return '/fostering/application'
        default:
            throw new Error('No matching endpoint for given stage.')
    }
}

const reduceProperties = object => Object.keys(object).reduce((acc, property) => {
    return {
        ...acc,
        [property]: object[property].value
    }
}, {})

const updateForm = async (endpoint, formData) => {
    const parsedFormData = parseFormData(formData)

    const response = await fetchWithTimeout(endpoint, {
        method: 'PATCH',
        credentials: 'include',
        body: JSON.stringify(parsedFormData),
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        }
    }, 30000)

    if (!response.ok) {
        throw Error(response.statusText)
    }

    return await response.json()
}

const updateFormStatus = ({ form, stage, currentStatus, setStatus }) => {
    if (currentStatus === TaskStatus.None) {
        fetchWithTimeout(`${getFormStageEndpoint(stage)}/status`,
            {
                method: 'PATCH',
                credentials: 'include',
                body: JSON.stringify({
                    status: TaskStatus.NotCompleted,
                    form
                }),
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                }
            }, 30000)

        setStatus(TaskStatus.NotCompleted)
    }
}

const parseFormData = ({ firstApplicant, secondApplicant, familyReference, firstPersonalReference, secondPersonalReference, ...formData }) => {
    let parsedObject = reduceProperties(formData)

    if (firstApplicant) {
        parsedObject.firstApplicant = reduceProperties(firstApplicant)
    }

    if (secondApplicant) {
        parsedObject.secondApplicant = reduceProperties(secondApplicant)
    }

    if (familyReference) {
        parsedObject.familyReference = reduceProperties(familyReference)
        parsedObject.familyReference.address = familyReference.address.value
    }

    if (firstPersonalReference) {
        parsedObject.firstPersonalReference = reduceProperties(firstPersonalReference)
        parsedObject.firstPersonalReference.address = firstPersonalReference.address.value
    }

    if (secondPersonalReference) {
        parsedObject.secondPersonalReference = reduceProperties(secondPersonalReference)
        parsedObject.secondPersonalReference.address = secondPersonalReference.address.value
    }

    return parsedObject
}

const updateApplicationForm = async (form, formData) => {
    const endpoint = getApplicationUpdateEndpoint(form)

    return await updateForm(endpoint, formData)
}

const updateHomeVisitForm = async (form, formData) => {
    const endpoint = getHomeVisitUpdateEndpoint(form)

    return await updateForm(endpoint, formData)
}

export {
    TaskStatus,
    StageName,
    HomeVisitFormName,
    ApplicationFormName,
    updateFormStatus,
    parseFormData,
    updateApplicationForm,
    updateHomeVisitForm
}