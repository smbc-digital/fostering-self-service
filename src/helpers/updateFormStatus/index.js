export const TaskStatus = {
    None: 0,
    Completed: 1,
    NotCompleted: 2,
    CantStart: 3
}

export const FormName =
{
    TellUsAboutYourself: 0,
    YourEmploymentDetails: 1,
    LanguageSpokenInYourHome: 2,
    YourPartnership: 3,
    YourFosteringHistory: 4,
    YourHealth: 5,
    TellUsAboutYourInterestInFostering: 6,
    YourHousehold: 7,
    ChildrenLivingAwayFromYourHome: 7
}

export default function (form, currentStatus, setStatus) {
    if (currentStatus === TaskStatus.None) {
        fetch('/front-end-endpoint-update-status',
            {
                method: 'PATCH',
                credentials: 'include',
                body: JSON.stringify({
                    status: TaskStatus.NotCompleted,
                    form
                })
            })

        setStatus(TaskStatus.NotCompleted)
    }
}