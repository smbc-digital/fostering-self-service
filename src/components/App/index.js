import React, { useContext, Fragment } from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'
import { Switch, Route, Redirect } from 'react-router'
import { getPageRoute } from 'helpers/pagehelper'
import { ErrorPage } from 'smbc-react-components'
import { Context } from 'context'
import Start from '../Pages/Start'
import { KnownByAnotherName, MoreAboutYou } from '../Pages/HomeVisit/1-TellUsAboutYou'
import { AreYouEmployed, EmploymentDetails } from '../Pages/HomeVisit/2-YourEmploymentDetails'
import LanguagesSpokenInYourHome from '../Pages/HomeVisit/3-LanguagesSpokenInYourHome'
import { AreYouMarried, MarriageDate, MovedInTogetherDate } from '../Pages/HomeVisit/4-PartnershipStatus'
import { HaveYouPreviouslyApplied } from '../Pages/HomeVisit/5-YourFosteringHistory'
import YourHealth from '../Pages/HomeVisit/6-YourHealth'
import { AboutYourInterest, TypesOfFostering } from '../Pages/HomeVisit/7-InterestInFostering'
import { AnyUnderSixteens, AboutUnderSixteens, AnyOverSixteens, AboutOverSixteens } from '../Pages/HomeVisit/8-ChildrenAwayFromHome'
import { AnyPeopleInYourHousehold, PeopleInYourHousehold, DoYouHaveAnyPets } from '../Pages/HomeVisit/10-YourHousehold'
import { YourGpDetails } from '../Pages/Application'

const JointApplicationOnlyRoute = ({ component: Component, ...props }) => {
	const { secondApplicant } = useContext(Context)

	return (
		<Route
			{...props}
			render={props =>
				secondApplicant
					? <Component {...props} />
					: <Redirect to={getPageRoute(1)} />}
		/>
	)
}

const App = () => {
	const { homeVisitDateTime } = useContext(Context)
	const disabledHomeVisitRoutes = moment().isSameOrAfter(moment(homeVisitDateTime.value, 'DD/MM/YYYY HH:mm').subtract(30, 'm'))

	return (
		<Switch>
			<Route exact path={getPageRoute(1)} component={Start} />
			<Route exact path="/error" component={ErrorPage} />
			{disabledHomeVisitRoutes || <Fragment>
				<Route exact path={`${getPageRoute(2)}/(|second-applicant)?`} component={KnownByAnotherName} />
				<Route exact path={`${getPageRoute(3)}/(|second-applicant)?`} component={MoreAboutYou} />
				<Route exact path={`${getPageRoute(4)}/(|second-applicant)?`} component={AreYouEmployed} />
				<Route exact path={`${getPageRoute(5)}/(|second-applicant)?`} component={EmploymentDetails} />
				<Route exact path={getPageRoute(6)} component={LanguagesSpokenInYourHome} />
				<JointApplicationOnlyRoute exact path={getPageRoute(7)} component={AreYouMarried} />
				<JointApplicationOnlyRoute exact path={getPageRoute(8)} component={MarriageDate} />
				<JointApplicationOnlyRoute exact path={getPageRoute(9)} component={MovedInTogetherDate} />
				<Route exact path={`${getPageRoute(10)}/(|second-applicant)?`} component={HaveYouPreviouslyApplied} />
				<Route exact path={`${getPageRoute(11)}/(|second-applicant)?`} component={YourHealth} />
				<Route exact path={getPageRoute(12)} component={AboutYourInterest} />
				<Route exact path={getPageRoute(13)} component={TypesOfFostering} />
				<Route exact path={getPageRoute(14)} component={AnyPeopleInYourHousehold} />
				<Route exact path={getPageRoute(15)} component={PeopleInYourHousehold} />
				<Route exact path={getPageRoute(16)} component={DoYouHaveAnyPets} />
				<Route exact path={`${getPageRoute(17)}/(|second-applicant)?`} component={AnyUnderSixteens} />
				<Route exact path={`${getPageRoute(18)}/(|second-applicant)?`} component={AboutUnderSixteens} />
				<Route exact path={`${getPageRoute(19)}/(|second-applicant)?`} component={AnyOverSixteens} />
				<Route exact path={`${getPageRoute(20)}/(|second-applicant)?`} component={AboutOverSixteens} />
				<Route exact path={`${getPageRoute(22)}/(|second-applicant)?`} component={YourGpDetails} />
			</Fragment>
			}
			<Redirect to={getPageRoute(1)} />
		</Switch>
	)
}

JointApplicationOnlyRoute.propTypes = {
	component: PropTypes.element
}

export default App