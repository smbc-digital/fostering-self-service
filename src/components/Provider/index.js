import React, { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import { Context } from '../../context/'
import { fetchWithTimeout } from '../../helpers'
import { API_ROOT } from '../../config'

export const Applicant = {
	FirstApplicant: 'firstApplicant',
	secondApplicant: 'secondApplicant'
}

const reduceProperties = object => Object.keys(object).reduce((acc, property) => {
	return {
		...acc,
		[property]: {
			value: object[property] || '',
			isValid: object[property] !== null
		}
	}
}, {})

const Provider = ({ children }) => {
	const displayRecaptcha = document.getElementById('displayRecaptcha') ? document.getElementById('displayRecaptcha').innerHTML === 'true' : false
	const initialState = {
		displayRecaptcha,
		currentApplicant: Applicant.FirstApplicant,
	}

	const [isLoading, setIsLoading] = useState(true)
	const [state, setState] = useState(initialState)
	const [error, setError] = useState(undefined)

	const onChange = (event, isValid) => {
		setState({
			...state,
			[event.target.name]: {
				value: event.target.value, 
				isValid
			}})
	}

	const onChangeApplicant = (event, isValid) => {
		setState({
			...state,
			[state.currentApplicant]: {
				...state[state.currentApplicant],
				[event.target.name]: {
					value: event.target.value, 
					isValid
				}
			}
		})
	}

	const setApplicant = applicant => {
		setState({
			...state,
			currentApplicant: applicant
		})
	}

	const mapCaseToContext = caseResponse => {
		const statuses = {...caseResponse.statuses}
		let secondApplicantDetails = undefined
		delete caseResponse.statuses

		const firstApplicantDetails = reduceProperties(caseResponse.firstApplicant)
		delete caseResponse.firstApplicant

		if (caseResponse.secondApplicant !== null) {
			secondApplicantDetails = reduceProperties(caseResponse.secondApplicant)
		}
		delete caseResponse.secondApplicant

		const caseDetails = reduceProperties(caseResponse)

		setState({
			...state, 
			...statuses, 
			firstApplicant: firstApplicantDetails, 
			secondApplicant: secondApplicantDetails, 
			...caseDetails
		})
	}

	const fetchCase = async () => {
		try {
			const response = await fetchWithTimeout(`${API_ROOT}/fostering/case`, { credentials: 'include' })
			const body = await response.json()
			mapCaseToContext(body)
		} catch (error) {
			setError(error)
		}
	}

	useEffect(() => {
		fetchCase()
	}, [])

	useEffect(() => {
		// Set isLoading to false only after the case is in state 
		if (isLoading && Object.keys(state).length > Object.keys(initialState).length) {
			setIsLoading(false)
		}
	}, [state])

	if (error) {
		return <p>Error</p>
	}

	if (isLoading) {
		return <p>Loading...</p>
	}

	return <Context.Provider value={{...state, onChange, onChangeApplicant, setApplicant}}>
		{children}
	</Context.Provider>
}

Provider.propTypes = {
	children: PropTypes.object
}

export default Provider