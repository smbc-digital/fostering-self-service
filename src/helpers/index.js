import fetchWithTimeout from './fetchWithTimeout'
import { getPageRoute } from './pagehelper'
import getCurrentApplicant from './getCurrentApplicant'
import { updateFormStatus, updateForm, FormName, TaskStatus } from './updateForm'

export {
    fetchWithTimeout,
    getPageRoute,
    getCurrentApplicant,
    updateFormStatus,
    updateForm,
    FormName,
    TaskStatus
}