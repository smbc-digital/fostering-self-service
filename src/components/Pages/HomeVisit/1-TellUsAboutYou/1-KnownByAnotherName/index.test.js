import { React, mount, useContextMock, renderer } from 'helpers/SetupTest'
import KnownByAnotherName from './index'
import { Applicant } from 'components/Provider'
import * as helpers from 'helpers'

describe('KnownByAnotherName', () => {

    const onChangeTargetMock = jest.fn()
    const onChangeStatusMock = jest.fn()

    beforeEach(() => {
        useContextMock.mockReturnValue({
            currentApplicant: Applicant.FirstApplicant,
            onChangeTarget: onChangeTargetMock,
            onChangeStatus: onChangeStatusMock,
            statuses: {
                TellUsAboutYourselfStatus: 0
            },
            firstApplicant: {
                everBeenKnownByAnotherName: {
                    value: 'true',
                    isValid: true
                },
                firstName: {
                    value: 'first name',
                    isValid: true
                },
                lastName: {
                    value: 'last name',
                    isValid: true
                },
                anotherName: {
                    value: 'last name',
                    isValid: true
                }
            },
            secondApplicant: {
                everBeenKnownByAnotherName: {
                    value: 'false',
                    isValid: true
                },
                firstName: {
                    value: 'second applicant first name',
                    isValid: true
                },
                lastName: {
                    value: 'second applicant last name',
                    isValid: true
                },
                anotherName: {
                    value: 'last name',
                    isValid: true
                }
            }
        })
    })
    
    it('should push to next page on submit, when first applicant', () => {
        // Arrange
        const history = {
            push: jest.fn()
        }

        const match = {
            params: undefined
        }

        const wrapper = mount(<KnownByAnotherName history={history} match={match}/>)

        // Act
        wrapper.find('Button').simulate('submit')

        // Assert
        expect(history.push).toHaveBeenCalled()
    })

    it('should push to next page on submit, when second applicant', () => {
        // Arrange
        const history = {
            push: jest.fn()
        }

        const match = {
            params: ['second-applicant']
        }

        const wrapper = mount(<KnownByAnotherName history={history} match={match}/>)

        // Act
        wrapper.find('Button').simulate('submit')

        // Assert
        const pageRoute = helpers.getPageRoute(3)
        expect(history.push).toHaveBeenCalledWith(pageRoute + '/second-applicant')
    })

    it('should call onChangeTarget', () => {
        // Arrange
        const history = {
            push: jest.fn()
        }

        const match = {
            params: undefined
        }

        const wrapper = mount(<KnownByAnotherName history={history} match={match}/>)

        // Act
        wrapper.find('input').first().simulate('change')

        // Assert
        expect(onChangeTargetMock).toHaveBeenCalled()
    })

    it('should update form status', async () => {
        // Arrange
        const history = {
            push: jest.fn()
        }

        const match = {
            params: undefined
        }
        
        const mockPromise = Promise.resolve()
        helpers.updateFormStatus = jest.fn().mockImplementation(({ setStatus, currentStatus }) => {
            setStatus(currentStatus)
        })
        helpers.fetchWithTimeout = jest.fn().mockReturnValue(mockPromise)

        // Act
        mount(<KnownByAnotherName history={history} match={match}/>)
        await mockPromise

        // Assert
        expect(onChangeStatusMock).toHaveBeenCalled()

    })

    describe('snapshot', () => {
        it('renders correctly', () => {
            const match = {
                params: undefined
            }

            const tree = renderer
                .create(<KnownByAnotherName history={{}} match={match}/>)
                .toJSON()

            expect(tree).toMatchSnapshot()
        })
    })
})