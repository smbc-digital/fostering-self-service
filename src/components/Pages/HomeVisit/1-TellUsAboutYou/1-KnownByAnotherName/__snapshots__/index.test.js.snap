// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KnownByAnotherName snapshot renders correctly 1`] = `
Array [
  <h1>
    Your fostering journey
  </h1>,
  <h2>
    Tell us more about you
  </h2>,
  <h3>
    first name
     
    last name
  </h3>,
  <form
    onSubmit={[Function]}
  >
    <h3
      className="no-margin-bottom"
    >
      Have you ever been known by another name?
    </h3>
    <fieldset>
      <div
        className="input-radio-container"
        role="radiogroup"
      >
        <input
          checked={true}
          disabled={undefined}
          id="have-you-been-known-yes"
          name="everBeenKnownByAnotherName"
          onChange={[Function]}
          type="radio"
          value="true"
        />
        <label
          htmlFor="have-you-been-known-yes"
        >
          Yes
        </label>
      </div>
      <div
        className="additional-list-component"
      >
        <label
          className="optional"
          htmlFor="anotherName"
        >
          Previous name
        </label>
        <input
          aria-label=""
          aria-live=""
          className=""
          id="anotherName"
          inputMode={undefined}
          maxLength="60"
          min={undefined}
          name="anotherName"
          onChange={[Function]}
          onKeyDown={[Function]}
          pattern={undefined}
          type="text"
          value="last name"
        />
      </div>
      <div
        className="input-radio-container"
        role="radiogroup"
      >
        <input
          checked={false}
          disabled={undefined}
          id="have-you-been-known-no"
          name="everBeenKnownByAnotherName"
          onChange={[Function]}
          type="radio"
          value="false"
        />
        <label
          htmlFor="have-you-been-known-no"
        >
          No
        </label>
      </div>
    </fieldset>
    <button
      className="button-primary"
      disabled={false}
      id={undefined}
      onClick={undefined}
      type="submit"
    >
      Next step
    </button>
    <p>
      <a
        href="#"
        onClick={[Function]}
      >
        Back
      </a>
    </p>
  </form>,
]
`;
